### ISSUE TITLE
NEMOH .h5 output problems

### ISSUE THREAD

**Issue by ahagmuller** [WEC-Sim/bemio#26]
When using our NEMOH single body .h5 file in WECSIM, the heave response is very strange as per attached image:

![image](https://cloud.githubusercontent.com/assets/13394872/9286805/28100556-42b1-11e5-83ac-23221aad3510.png)

When I change the simulation time from the 150 Seconds above to only 15 seconds I get this:

![image](https://cloud.githubusercontent.com/assets/13394872/9286837/3967b096-42b2-11e5-8912-a2b4388c89fd.png)

At ten seconds:

![image](https://cloud.githubusercontent.com/assets/13394872/9286852/9ae3d4f8-42b2-11e5-9b19-64da63d1fd28.png)

Not sure what to make of it, but by the scale of the graph, it appears to be the same result just zoomed in for the 150 second simulation. Something similar happened for the NEMOH single body tutorial file I used as well  per below:

 
![image](https://cloud.githubusercontent.com/assets/13394872/9286779/87e86f64-42b0-11e5-8a64-9a9260eae277.png)

Basically, it sent the point absorber to Lagrange point 5 by the end of the simulation. Our wave energy device seems more useful as a spaceship.

**Comment by bjrose86** [comment]
I had the same issue, which i think is a numerical instability. Try lowering the simulation dt. It will take much longer to run, but should prevent the ODE solver from going crazy

**Comment by ahagmuller** [comment]
I changed the simulation dt to .01 from .1 and this is the result with a 200 second simulation time:
![image](https://cloud.githubusercontent.com/assets/13394872/9287842/aa92029e-42df-11e5-8521-d3ca55ab685d.png)

Did not seem to fix the issue, I will try a smaller dt, but this took 746 seconds to run..

**Comment by ahagmuller** [comment]
I tried to model the rm3 body 1 float from wamit output in nemoh to compare data, as well as for the body that we are working with in nemoh. In the screenshot below, the left is the added mass for the nemoh outputs and the right for the wamit outputs. Things seem orders of magnitude different. 

![image](https://cloud.githubusercontent.com/assets/13394872/9290057/612eaf22-4338-11e5-9c21-2a0053877450.png)

**Comment by sabible** [comment]
I've been noticing similar things.  

First, the hydrostatic linear stiffness coefficients are many order of magnitude different for the WEC-SIM tutorial H5s using Wamit data and data I produce through NEMOH for the same geometry.  Bemio is reading the data from Nemoh correctly, so this is an issue of NEMOH and WAMIT producing different data.  Does anyone know of another way to produce the hydrostatic stiffness matrix to check which one is right?

Second, for the excitation, added mass, etc..  The values are not so much different (but a little) but there is a read-in problem on Bemio.  Notice that for the Wamit data, the values in the H5 rows are for each of the 6 forces.  Moving left-to-right, each column is representing each of the wave directions simulated.  For NEMOH data Bemio is doing the exact opposite, using 6 columns for the forces and each row is representing a new wave direction.  When you try to import more than 6 wave directions you get an error about matrix size.

**Comment by lawsonro3** [comment]
I just took a look at this problem and made a small change that I think will fix this problem.

Spring stiffness scaling - I fixed the scaling of the in the Nemoh data reader - it was indeed incorrect in that it was not being scaled

Ordering of Nemoh coeffs - I believe this was fixed a while ago, please pull the latest version of the code and try again.

*****Also***** make sure that in the `scale=False` is specified as an argument to wamit.read or nemoh.read functions. WEC-Sim expects the `scale=False`. If you do not specify the `scale` the default is false. The WEC-Sim and bemio documentation is being updated to reflect this as we speak.

Please let me know if you are still having problems.

-Mike

**Comment by lawsonro3** [comment]
Please let me know if this problem is not fixed.

**Comment by ahagmuller** [comment]
Hi Mike,

I Pulled the latest code and ran again and still get the same strange results. As far as I can tell, any NEMOH generated file that I run has the same kind of strange results. One other thing I am noticing, and I am not sure if this is even related, is that I have problems with the IRF time being less than the CIC time. When I reduce the CIC time this error goes away, but I am not sure why this is a problem to begin with. I will open a new issue for this, as I am not certain they are even related.

At this point, I have generated a NEMOH mesh file based on the geometry of the float for the RM3 tutorial file, and will post results for both in this issue.

Thanks,

Alex

**Comment by lawsonro3** [comment]
Alex, could you post files so I can test the example that is giving you the problem?

-Mike

**Comment by ahagmuller** [comment]
I will get this to you by Monday. Do you want the raw NEMOH outputs, the
generated .h5 files, the simulink file, and input files?

Alex

**Comment by lawsonro3** [comment]
Everything if possible.

Thanks!

-Mike

**Comment by CWindt** [comment]
Hi Alex,

I'm pretty keen on rerunning your case, since I'm also struggling with the .h5 file from Nemoh. So if possible, can you also give me access to your files?

Cheers,
Christian

**Comment by ahagmuller** [comment]
Christian,

I will try and get these posted this evening.

Regards,
Alex

**Comment by lawsonro3** [comment]
Hi Cristian,

If you send me your files, I could debug it from there too. The cases I have are working, so I'm guessing its a small bug that could be fixed rather quickly if I have the Nemoh files.

Best,

-Mike

**Comment by ahagmuller** [comment]
Mike/Christian,
I uploaded the files here: https://github.com/ahagmuller/WEC-SIM-AquaHarmonics_Shared

To make a long story short, we discovered that the computer we were working with had a failing hard drive, and it's memory appeared "swiss cheesed". We replaced the hard drive and started with a clean install of everything. We worked from scratch with the analysis starting with using aximesh.m to generate a mesh file for NEMOH based on the RM3 buoy "A" geometry. I ran into a few small snags but worked through them and I documented the findings in the "readme" pdf in the repository. 

Ultimately, everything worked, and I can't say that I did anything different from before, so it is possible that something in the Python install was not correct, or the failing hard drive had something to do with it.

While this appears successful, and a comparison of some of the outputs for buoy"A" of the RM3 reference model analyzed in WEC SIM with the NEMOH and WAMIT generated .h5 files are shown in the readme.pdf file.

I do have a few questions however when using NEMOH in general:

The CIC time is set to 60 seconds as default, so when running NEMOH, I hav
e been using 200 seconds as the IRF time with a step of .1 which is in exce
ss of the 60 seconds CIC time, and there does not appear to be any issues h
ere. What would be the minimum IRF time and step size here? Do you want you
r IRF time to be as long as the duration of the simulation time in WECSIM, 
with a similar timestep?

-When calculating IRF in BEMIO from the run.py file, do you want the endtim
e value to be the same as the IRF duration in the nemoh.cal file, or does t
his even matter?

-NEMOH does not directly calculate zero frequency added mass, and the worka
round used is to use a very small frequency here that i found in NEMOH foru
ms I have read. I used 0.01 as my starting frequency (in rad/s) for the IRF
calculation. Does WEC SIM even need or use the zero frequency added mass f
or any calculations? Do you just need to have the range of frequencies cove
red that you would be testing in WECSIM?

Please let me know if you have any questions as well.
I can post my above questions elsewhere if that is more appropriate.

Thanks,

Alex

**Comment by lawsonro3** [comment]
Hi Alex,

Thanks for the feedback and I'm glad that you were able to get WEC-Sim/bemio working using Nemoh.

In fact, its great to see the comparison between the standard WAMIT RM3 tutorial and the Nemoh RM3 case you ran. Nice work! Would it be ok with you if you to license you repository - https://github.com/ahagmuller/WEC-SIM-AquaHarmonics_Shared - under an Apache 2.0 so we could include some of it into the bemio repo and so others can benefit from your work?

Answers to a couple of your other questions:
* CIC time - in the bemio calc_irf_radiation and calc_irf_excitation be sure to select a t_end to a number where the IRD decays appropriately to 0. Selecting a time that is too long will results in unnecessary data  being stored in the .h5 file, while selecting a t_end that is too short will results in the IRF being cut off too soon. In WEC-Sim the CIC time must be set equal to or less than the IRF t_end time used in bemio for obvious reasons. I believe that WEC-Sim automatically checks this and throws an error if CIC time > bemio t_end.
* *When calculating IRF in BEMIO from the run.py file, do you want the endtim
e value to be the same as the IRF duration in the nemoh.cal file, or does this even matter?* Answer: It does not matter because bemio does not read the NEMOH value.
* Inf freq added mass - NEMOH does indeed provide the infinite frequency added mass term in its output files and bemio will read this data. To enable calculation of the inf freq added mass in Nemoh, you must enable the IRF option is set to 1 in the Nemoh.cal file. NEMOH will then produce the IRF.tec files, which contains that inf freq added mass data. Note that bemio does not read the IRF data from the IRF.tec file, only the inf freq added mass is read.

If any of this is unclear please let me know and I can provide some more detailed info.

-Mike

**Comment by lawsonro3** [comment]
Cristian - does my previous post address your question as well?

**Comment by CWindt** [comment]
Alex, thanks for the data. I will try to set up these benchmarks on my machine an see how that goes. So Mike, so far I'm happy with the answers you gave here. If I face further problems while running the cases, I will get back to you. But thanks anyways!
Christian

**Comment by ahagmuller** [comment]
Hi Mike,
Most of your answers are clear, I think the only one I am still confused on
was the added mass... I may have posted infinite frequency, but I meant
zero frequency added mass for NEMOH....
That is not directly calculated in the.tec file, but can be computed by
choosing a very small frequency to start with in the IRF calculation in
NEMOH...
So my question then is, does WEC-SIM need/use zero frequency added mass?

Thanks!

Also,

I will setup the Apache licence soon, I should have done it when I started
the repository, but I think I can just add it as a file.

 Also, it sounded like the issues I documented in my readme.pdf about the
need to eliminate the 3 dashes and ensure there are only 34 lines in the
nemoh.cal file prior to running BEMIO were also discovered by the  group
who was having problems running with less than 6 DOF (ISSUE #19).

Best,

Alex

**Comment by ahagmuller** [comment]
Hi Christian,

I would also suggest running axisym from scratch if you have not for the
mesh at some point to ensure all vectors are pointing "outward" from your
geometry, as shown in the readme.pdf. If you setup the points in the wrong
direction, they will point "inward" and although NEMOH will run, it will
cause errors in WEC-SIM.

Regards,

Alex

**Comment by lawsonro3** [comment]
Hi Alex,

WEC-Sim never uses/needs zero freq added mass. Only the infinite frequency as described in the WEC-Sim users manual - http://wec-sim.github.io/WEC-Sim/.

Thanks for agreeing to put in the Apache license files. You are right all you have to do is add then in to your repo.

Thanks again!

-Mike

**Comment by ahagmuller** [comment]
Hi Mike,
I added the Apache License file to the repository.
Is there anything else needed there?
Thanks,
Alex

**Comment by lawsonro3** [comment]
Nope, I think that does it. Thanks for all the help and let me know if you run into any other problems.

-Mike

**Comment by Owen0708** [comment]
Hi,
I have two question:
1. I have the same problem as the above graph, like this:
![image](https://cloud.githubusercontent.com/assets/13230160/12603185/4b6799e4-c4eb-11e5-89bf-2f833bebd197.png)
so I am try to change my CIC time, could anybody kindly help me, how to select a t_end to a number where the IRD decays appropriately to 0 ?
I am confuse in there relationship between CIC time & IRF.
and here is my input in Nemoh.cal & Bemio run.py.
Nemoh.cal: ![image](https://cloud.githubusercontent.com/assets/13230160/12603013/697ef582-c4e9-11e5-84b6-70613d51d149.png)
run.py: 
![image](https://cloud.githubusercontent.com/assets/13230160/12603020/857b4114-c4e9-11e5-9732-ddcb004b622b.png)
2. another question is : if I set the wave frequency in a range between 1 and 10 in the Nemoh program, the WECSim input file "wecSimInputFile" may I set a wave period exceed the range from the Nemoh?  

Can anyone help me? I would appreciate it if you could help me.
owen
